extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -16,45 +16,175 @@
  */
-class AvroSerializerHooks {
+public final class AvroSerializerHooks implements DataSerializerHook {
+    public static final int F_ID = FactoryIdHelper.getFactoryId(AVRO_SCHEMA_DS_FACTORY, AVRO_SCHEMA_DS_FACTORY_ID);
+    public static final int SCHEMA = 1;
 
-    public static final class Utf8Hook implements SerializerHook<Utf8> {
+    private static final int SCHEMA_ID_SIZE = 8;
+
+    @Override
+    public int getFactoryId() {
+        return F_ID;
+    }
+
+    @Override
+    public DataSerializableFactory createFactory() {
+        return typeId -> typeId == SCHEMA ? new AvroSchema() : null;
+    }
+
+    public static <T> byte[] serialize(DatumWriter<T> datumWriter, T data) {
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+            serialize(out, datumWriter, data);
+            return out.toByteArray();
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static <T> void serialize(ByteArrayOutputStream out, DatumWriter<T> datumWriter, T data) throws IOException {
+        BinaryEncoder encoder = EncoderFactory.get().directBinaryEncoder(out, null);
+        datumWriter.write(data, encoder);
+        encoder.flush();
+    }
+
+    public static <T> T deserialize(DatumReader<T> datumReader, byte[] data) {
+        return deserialize(datumReader, data, 0, data.length);
+    }
+
+    public static <T> T deserialize(DatumReader<T> datumReader, byte[] data, int offset, int length) {
+        try {
+            BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(data, offset, length, null);
+            return datumReader.read(null, decoder);
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static class GenericRecordHook implements SerializerHook<GenericRecord>, HazelcastInstanceAware {
+        private final Map<Schema, AvroSchema> replicatedSchemas = new HashMap<>();
+        private SchemaService schemaService;
 
+        @Override
+        public Class<GenericRecord> getSerializationType() {
+            return GenericRecord.class;
+        }
+
+        @Override
+        public Serializer createSerializer() {
+            return new ByteArraySerializer<GenericRecord>() {
+                @Override
+                public int getTypeId() {
+                    return SerializerHookConstants.AVRO_GENERIC_RECORD;
+                }
+
+                @Override
+                public byte[] write(GenericRecord record) {
+                    if (record instanceof LazyImmutableRecord) {
+                        return ((LazyImmutableRecord) record).serializedRecord;
+                    }
+                    try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+                        AvroSchema schema = replicatedSchemas.get(record.getSchema());
+                        if (schema == null) {
+                            schema = new AvroSchema(record.getSchema());
```

Comment:
what is the benefit of using cache here? Not having to calculate Rabin fingerprint of the schema (is hashCode + hashmap lookup + equals cheaper)? `SchemaService` has local cache of all schemas